### PR TITLE
Add Gaussian NLL loss function

### PR DESCRIPTION
**Pull request by AtomicVar**
## Proposed changes

This PR adds the [Gaussian negative log likelihood](https://pytorch.org/docs/stable/generated/torch.nn.GaussianNLLLoss.html) loss function. It is frequently used in situations where you have a probabilistic model that predicts a distribution (usually Gaussian or normal distribution) as the output.

The interface and implementation are referenced from PyTorch. And the ground-truth values in the testing code are calculated by PyTorch.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: python/mlx/nn/losses.py [hunk 1]
```diff
@@ -169,6 +169,54 @@ def nll_loss(
...
+
+    Args:
+        inputs (array): The predicted expectation of the Gaussian distribution.
+        targets (array): The target values (samples from the Gaussian distribution).
+        vars (array): The predicted variance of the Gaussian distribution.
+        full (bool, optional): Whether to include the constant term in the loss calculation.
+            Default: ``False``.
+        eps (float, optional): Small positive constant to prevent numerical instability.
+            Defaults to ``1e-6``.
+
```
Comment (ligaz) [inline python/mlx/nn/losses.py:196]: Can you add description for the `reduction` argument?

File: python/mlx/nn/losses.py [hunk 2]
```diff
@@ -169,6 +169,54 @@ def nll_loss(
...
+        raise ValueError(
+            f"Inputs shape {inputs.shape} does not match targets shape {targets.shape}."
+        )
+
+    if inputs.shape != vars.shape:
+        raise ValueError(
+            f"Inputs shape {inputs.shape} does not match vars shape {vars.shape}."
+        )
+
+    loss = 0.5 * mx.log(mx.maximum(vars, eps)) + 0.5 * mx.square(
```
[the remark this answers is not in the document]
Comment (AtomicVar) [reply python/mlx/nn/losses.py]: Thanks for your advice. I refactored this.

File: python/mlx/nn/losses.py [hunk 3]
```diff
@@ -169,6 +169,54 @@ def nll_loss(
     return _reduce(loss, reduction)
 
 
+def gaussian_nll_loss(
+    inputs: mx.array,
+    targets: mx.array,
+    vars: mx.array,
+    full: bool = False,
+    eps: float = 1e-6,
+    reduction: str = "none",
```
[the remark this answers is not in the document]
Comment (AtomicVar) [reply python/mlx/nn/losses.py]: That's nice. I refactored my code and push a few more commits according to your advice. Can you check them again? @ligaz Thanks!

File: python/mlx/nn/losses.py (hunk 1 above)
Comment (AtomicVar) [reply python/mlx/nn/losses.py:200]: Done.

File: python/mlx/nn/losses.py [hunk 4]
```diff
@@ -1,12 +1,14 @@
 # Copyright © 2023 Apple Inc.
 
 import math
+from typing import Literal, TypeAlias
 
 import mlx.core as mx
-from mlx.nn.layers.base import Module
 
+Reduction: TypeAlias = Literal["none", "mean", "sum"]
```
Comment (awni) [inline python/mlx/nn/losses.py:8]: This is really nice! And sadly it's 3.11 and up. So do you mind removing it since we currently support 3.8 and up?

**Review by awni — CHANGES_REQUESTED** [review]
This is nicely done! Thanks for adding it! I wish we could keep the `Reduction` `TypeAlias` but it won't work < 3.11 so would you mind removing those changes?

O/w we can get this merged soon!

File: python/mlx/nn/losses.py (hunk 4 above)
Comment (AtomicVar) [reply python/mlx/nn/losses.py]: I removed the `TypeAlias` and tested it in a Python 3.8 environment. Is it OK now? 😃

**Review by awni — APPROVED** [review]
Thanks, looks great!!
